gui/biomes.lua

```diff
@@ -0,0 +1,213 @@
+    DESERT_SAND = "Sand Desert",
+    OCEAN_TROPICAL = "Tropical Ocean",
+    OCEAN_TEMPERATE = "Temperate Ocean",
+    OCEAN_ARCTIC = "Arctic Ocean",
+    POOL_TEMPERATE_FRESHWATER = "Temperate Freshwater Pool",
+    POOL_TEMPERATE_BRACKISHWATER = "Temperate Brackishwater Pool",
+    POOL_TEMPERATE_SALTWATER = "Temperate Saltwater Pool",
+    POOL_TROPICAL_FRESHWATER = "Tropical Freshwater Pool",
+    POOL_TROPICAL_BRACKISHWATER = "Tropical Brackishwater Pool",
+    POOL_TROPICAL_SALTWATER = "Tropical Saltwater Pool",
+    LAKE_TEMPERATE_FRESHWATER = "Temperate Freshwater Lake",
+    LAKE_TEMPERATE_BRACKISHWATER = "Temperate Brackishwater Lake",
+    LAKE_TEMPERATE_SALTWATER = "Temperate Saltwater Lake",
+    LAKE_TROPICAL_FRESHWATER = "Tropical Freshwater Lake",
+    LAKE_TROPICAL_BRACKISHWATER = "Tropical Brackishwater Lake",
+    LAKE_TROPICAL_SALTWATER = "Tropical Saltwater Lake",
+    RIVER_TEMPERATE_FRESHWATER = "Temperate Freshwater River",
+    RIVER_TEMPERATE_BRACKISHWATER = "Temperate Brackishwater River",
+    RIVER_TEMPERATE_SALTWATER = "Temperate Saltwater River",
+    RIVER_TROPICAL_FRESHWATER = "Tropical Freshwater River",
+    RIVER_TROPICAL_BRACKISHWATER = "Tropical Brackishwater River",
+    RIVER_TROPICAL_SALTWATER = "Tropical Saltwater River",
+    SUBTERRANEAN_WATER = "Subterranean Water",
+    SUBTERRANEAN_CHASM = "Subterranean Chasm",
+    SUBTERRANEAN_LAVA = "Subterranean Lava",
+}
+
+local function find(t, predicate)
+    for k, item in pairs(t) do
+        if predicate(k, item) then
+            return k, item
+        end
+    end
+    return nil
+end
+
+local biomesMap = {}
+local biomeList = {}
+local function gatherBiomeInfo()
+    local maxX, maxY, maxZ = dfhack.maps.getTileSize()
+    maxX = maxX - 1; maxY = maxY - 1; maxZ = maxZ - 1
+
+    local z = df.global.window_z
+
+    for y = 0, maxY do
+        for x = 0, maxX do
+            local rgnX, rgnY = dfhack.maps.getTileBiomeRgn(x,y,z)
+            local biomeTypeId = dfhack.maps.getBiomeType(rgnX, rgnY)
+            local biome = dfhack.maps.getRegionBiome(rgnX, rgnY)
+
+            local biomesZ = biomesMap[z]
+            if not biomesZ then
+                biomesZ = {}
+                biomesMap[z] = biomesZ
+            end
+            local biomesZY = biomesZ[y]
+            if not biomesZY then
+                biomesZY = {}
+                biomesZ[y] = biomesZY
+            end
+
+            local function currentBiome(_, item)
+                return item.biome == biome
+            end
+            local ix = find(biomeList, currentBiome)
+            if not ix then
+                local ch = string.char(string.byte('a') + #biomeList)
+                table.insert(biomeList, {biome = biome, char = ch, typeId = biomeTypeId})
+                ix = #biomeList
+            end
+
+            biomesZY[x] = ix
+        end
+    end
+end
+
+gatherBiomeInfo()
+
+local TITLE = "Biomes"
+
+BiomeVisualizerLegend = nil -- for debugging purposes
```

Comment:
for clearing the script environment for debugging/dev purposes, consider using `devel/clear-script-env`

Reply:
I probably don't understand this. Right now, if I make a change in the script, or more specifically in the `BiomeVisualizerLegend`-class, I execute `gui/biomes`, and the changes are immediately applied. How would I use `devel/clear-script-env` instead? Execute first the `clear-sript-env` and then `gui/biomes`?.. I don't remember, how can I chain commands in dfhack console?

Reply:
you'd only need to use `devel/clear-script-env` if you had previously set a global in the script and an update removes the global. for example, if you had `function MyClass:onInput(keys)` but then you remove the function, *it will still be defined and active* until you run `devel/clear-script-env` (or reload the game). If you just change the function definition, then there's no problem.

Reply:
well, then I have to think about whether I set a global or not, and since most of the time I do not, I'll definitely forget the one time I do.